src/Microsoft.AspNetCore.Routing/RouteValuesBasedEndpointFinder.cs

```diff
@@ -0,0 +1,191 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Routing.EndpointFinders;
+using Microsoft.AspNetCore.Routing.Internal;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.AspNetCore.Routing.Template;
+using Microsoft.AspNetCore.Routing.Tree;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class RouteValuesBasedEndpointFinder : IEndpointFinder<RouteValuesBasedEndpointFinderContext>
+    {
+        private readonly CompositeEndpointDataSource _endpointDataSource;
+        private readonly IInlineConstraintResolver _inlineConstraintResolver;
+        private readonly ObjectPool<UriBuildingContext> _objectPool;
+        private LinkGenerationDecisionTree _allMatchesLinkGenerationTree;
+        private IDictionary<string, NamedMatch> _namedMatches;
+
+        public RouteValuesBasedEndpointFinder(
+            CompositeEndpointDataSource endpointDataSource,
+            ObjectPool<UriBuildingContext> objectPool,
+            IInlineConstraintResolver inlineConstraintResolver)
+        {
+            _endpointDataSource = endpointDataSource;
+            _objectPool = objectPool;
+            _inlineConstraintResolver = inlineConstraintResolver;
+
+            BuildOutboundMatches();
+        }
+
+        public IEnumerable<Endpoint> FindEndpoints(RouteValuesBasedEndpointFinderContext context)
+        {
+            IEnumerable<OutboundMatchResult> matchResults = null;
+            if (string.IsNullOrEmpty(context.RouteName))
+            {
+                matchResults = _allMatchesLinkGenerationTree.GetMatches(
+                    context.ExplicitValues,
+                    context.AmbientValues);
+            }
+            else if (_namedMatches.TryGetValue(context.RouteName, out var namedMatch))
+            {
+                // Since a route name could be associated with multiple endpoints, check to see if we need the
+                // decision tree to give us the matches or not.
+                if (namedMatch.Match == null)
+                {
+                    matchResults = namedMatch.LinkGenerationTree.GetMatches(
+                        context.ExplicitValues,
+                        context.AmbientValues);
+                }
+                else
+                {
+                    // When the route name is associated with a single endpoint
+                    return new[] { (MatcherEndpoint)namedMatch.Match.Entry.Data };
+                }
+            }
+
+            if (matchResults == null || !matchResults.Any())
+            {
+                return null;
+            }
```

Comment:
return `Array.Empty<>` here. Returning a null IEnumerable is a party foul, no one expect you to do that